src/providers/postgres/qgspostgresproviderconnection.cpp

```diff
@@ -656,3 +656,48 @@ QList<QgsVectorDataProvider::NativeType> QgsPostgresProviderConnection::nativeTy
   }
   return types;
 }
+
+
+QgsFields QgsPostgresProviderConnection::fields( const QString &schema, const QString &tableName ) const
+{
+  // Try the base implementation first and fall back to a more complex approch for the
+  // few PG-specific corner cases that do not work with the base implementation.
+  try
+  {
+    return QgsAbstractDatabaseProviderConnection::fields( schema, tableName );
+  }
+  catch ( QgsProviderConnectionException &ex )
+  {
+    // This table might expose multiple geometry columns (different geom type or SRID)
+    // but we are only interested in fields here, so let's pick the first one.
+    TableProperty tableInfo { table( schema, tableName ) };
+    if ( tableInfo.geometryColumnTypes().count( ) > 1 )
+    {
+      try
+      {
+        QgsDataSourceUri tUri { tableUri( schema, tableName ) };
+        TableProperty::GeometryColumnType geomCol { tableInfo.geometryColumnTypes().first() };
+        tUri.setGeometryColumn( tableInfo.geometryColumn() );
+        tUri.setWkbType( geomCol.wkbType );
+        tUri.setSrid( QString::number( geomCol.crs.postgisSrid() ) );
+        if ( tableInfo.primaryKeyColumns().count() > 0 )
+        {
+          tUri.setKeyColumn( tableInfo.primaryKeyColumns().first() );
+        }
+        tUri.setParam( QStringLiteral( "checkPrimaryKeyUnicity" ), QLatin1String( "0" ) );
+        QgsVectorLayer::LayerOptions options { true, true };
```

Comment:
```suggestion
        QgsVectorLayer::LayerOptions options { false, false };
```

Reply:
Why `readExtentFromXml`? I'm only interested in the fields here, we don't need any extent or style information, I thought that by constructing the options with `readExtentFromXml`=`false` we can speed things up, you are right about `loadDefaultStyle` though, I think it skips some loading operations.

Reply:
Ok good point -- readExtentFromXml SHOULD be true!